src/nodes.coffee

```diff
@@ -2336,13 +2348,37 @@ exports.Range = class Range extends Base
     idx      = del o, 'index'
     idxName  = del o, 'name'
     namedIndex = idxName and idxName isnt idx
-    varPart  =
-      if known and not namedIndex
-        "var #{idx} = #{@fromC}"
+    if o.declare  # from compileArray: need to declare all variables as local
+      declare = yes
+    else
+      # Can do var declaration only if all relevant variables are in this scope.
+      # If we declare @fromVar, @toVar, or @stepVar, they're guaranteed
+      # to be in scope (generated in @compileVariables), so don't check those.
```

Comment:
The comments become the annotated source, so please always use curly quotes and backticks, and proper English:

```suggestion
      # We can output `var` declarations only if all relevant variables are in this scope.
      # If we declare `@fromVar`, `@toVar`, or `@stepVar`, they’re guaranteed
      # to be in scope (generated in `@compileVariables`), so don’t check those.
```

And so on for all the comments. The logic you’re adding in this PR is quite difficult to follow (at least by me) and excessive documentation of what it’s doing would be really good to have.